src/pinnable_mapped_file.cpp

```diff
@@ -258,29 +376,52 @@ bool pinnable_mapped_file::all_zeros(char* data, size_t sz) {
    return true;
 }
 
-void pinnable_mapped_file::save_database_file() {
-   std::cerr << "CHAINBASE: Writing \"" << _database_name << "\" database file, this could take a moment..." << std::endl;
-   char* src = (char*)_non_file_mapped_mapping;
-   char* dst = (char*)_file_mapped_region.get_address();
+std::pair<std::byte*, size_t> pinnable_mapped_file::get_region_to_save() const {
+   if (_non_file_mapped_mapping)
+      return { (std::byte*)_non_file_mapped_mapping, _database_size };
+   return { (std::byte*)_file_mapped_region.get_address(), _database_size };
+}
+
+void pinnable_mapped_file::save_database_file(bool flush /* = true */) {
+   assert(_writable);
+   std::cerr << "CHAINBASE: Writing \"" << _database_name << "\" database file, this could take a moment..." << '\n';
    size_t offset = 0;
    time_t t = time(nullptr);
-   while(offset != _file_mapped_region.get_size()) {
-      if(!all_zeros(src+offset, _db_size_multiple_requirement))
-         memcpy(dst+offset, src+offset, _db_size_multiple_requirement);
-      offset += _db_size_multiple_requirement;
+   pagemap_accessor pagemap;
+   size_t written_pages {0};
+   auto [src, sz] = get_region_to_save();
+   
+   while(offset != sz) {
+      size_t copy_size = std::min(_db_size_copy_increment,  sz - offset);
+      bool mapped_writable_instance = std::find(_instance_tracker.begin(), _instance_tracker.end(), this) != _instance_tracker.end();
+      if (!mapped_writable_instance ||
+          !pagemap.update_file_from_region({ src + offset, copy_size }, _file_mapping, offset, flush, written_pages)) {
+         if (mapped_writable_instance)
+            std::cerr << "CHAINBASE: ERROR: pagemap update of db file failed... using non-pagemap version" << '\n';
+         if(!all_zeros(src+offset, copy_size)) {
+            bip::mapped_region dst_rgn(_file_mapping, bip::read_write, offset, copy_size);
+            char *dst = (char *)dst_rgn.get_address();
+            memcpy(dst, src+offset, copy_size);
+
+            if (flush) {
+               std::cerr << "CHAINBASE: Writing \"" << _database_name << "\" database file, flushing buffers..." << '\n';
```

Comment:
I don't think this message makes sense to move in to the while loop. Previously it was only printed once: at the end. Now it's printed for _every_ (non-zero) 1GB of memory -- it can even be printed more times than the status message if you're sinking more than 1GB/sec.

I think you can just get rid of this message entirely. Its original purpose was to indicate a final end stage that may not be accounted for in the percentage status messages. But if syncing every 1GB, that isn't needed any longer.

Reply:
Yes, I agree.

Reply:
done.